src/H5Z.c

```diff
@@ -581,26 +603,51 @@ H5Z__check_unregister_dset_cb(void *obj_ptr, hid_t H5_ATTR_UNUSED obj_id, void *
  *-------------------------------------------------------------------------
  */
 static int
-H5Z__flush_file_cb(void *obj_ptr, hid_t H5_ATTR_UNUSED obj_id, void H5_ATTR_PARALLEL_USED *key)
+H5Z__flush_file_cb(void H5_ATTR_UNUSED *obj_ptr, hid_t H5_ATTR_UNUSED obj_id, void H5_ATTR_PARALLEL_USED *key)
 {
-    H5F_t *f = (H5F_t *)obj_ptr; /* File object for operations */
+
 #ifdef H5_HAVE_PARALLEL
     H5Z_object_t *object = (H5Z_object_t *)key;
-#endif                     /* H5_HAVE_PARALLEL */
-    int ret_value = false; /* Return value */
+#endif                                              /* H5_HAVE_PARALLEL */
+    int                       ret_value = false;    /* Return value */
+    H5VL_file_specific_args_t vol_cb_args_specific; /* Arguments to VOL callback */
+    H5VL_object_t            *vol_obj;              /* File for file_id */
+    H5VL_file_get_args_t      vol_cb_args;          /* Arguments to VOL callback */
+
+    unsigned int intent = 0;
 
     FUNC_ENTER_PACKAGE
 
     /* Sanity checks */
-    assert(obj_ptr);
     assert(key);
 
+    /* Get the internal file structure */
+    if (NULL == (vol_obj = (H5VL_object_t *)H5I_object(obj_id)))
+        HGOTO_ERROR(H5E_ARGS, H5E_BADTYPE, FAIL, "invalid file identifier");
+
+    /* Get intent */
+    vol_cb_args.op_type               = H5VL_FILE_GET_INTENT;
+    vol_cb_args.args.get_intent.flags = &intent;
+
+    /* Get the flags */
+    if (H5VL_file_get(vol_obj, &vol_cb_args, H5P_DATASET_XFER_DEFAULT, H5_REQUEST_NULL) < 0)
+        HGOTO_ERROR(H5E_FILE, H5E_CANTGET, FAIL, "unable to get file's intent flags");
+
     /* Do a global flush if the file is opened for write */
-    if (H5F_ACC_RDWR & H5F_INTENT(f)) {
+    if (H5F_ACC_RDWR & intent) {
 
 #ifdef H5_HAVE_PARALLEL
+        H5G_loc_t loc;
+        H5F_t    *f = NULL;
+
         /* Check if MPIO driver is used */
+        if (H5G_loc(obj_id, &loc) < 0)
+            HGOTO_ERROR(H5E_ARGS, H5E_BADTYPE, FAIL, "not a location");
+        f = loc.oloc->file;
```

Comment:
This appears to have the same issue of assuming that we can cast things to an `H5F_t` and perform native operations on them

Reply:
You're right, `H5G_loc` does the same thing under the hood. Even if we didn't use it here, `H5F_mpi_retrieve_comm` invokes `H5G_loc` on the same object id anyway. Would it make more sense to just throw an error if we're using a non-native VOL and HDF5 was built parallel?

Reply:
Since the filter interface is not exactly exposed to non-native VOLs other than having the ability to make H5Z calls, I would probably check if the object is native and just skip this whole block if it's not, but don't throw an error. I would have done the same for the rest of the callbacks changed in this PR - just continue iteration if the object is non-native - but this PR is a good change towards being able to have non-native VOLs interact with the filter interface in the future.